### DISCUSSION TITLE
Downloading 5 tweets and rate limit immediately reached

### DISCUSSION THREAD

**Discussion by John-Kaz** [cjbarrie/academictwitteR#195] [Q&A]
A script using a for loop to run `get_user_tweets` one user at a time downloaded over a thousand users' 1 million tweets a couple of weeks ago. I updated the package and tried using `get_all_tweets`, again, one user at a time. This time, however, I get called on hitting the data limit almost immediately and the script keeps waiting for 15 minute intervals. For instance, the first time I tried today, the first user's 5 tweets downloaded just fine but when it attempted downloading the second user, I got hit with a 15 minute wait period. I have a developer Twitter account and bearer, and am nowhere near my 10 million monthly quota. Very befuddled...

**Comment by chainsawriot** [comment]
As your question is not an issue (e.g. bug report or issue request), I moved this here.

**Reply by shmuhammadd** [reply]
> A script using a for loop to run `get_user_tweets` one user at a time downloaded over a thousand users' 1 million tweets a couple of weeks ago. I updated the package and tried using `get_all_tweets`, again, one user at a time. This time, however, I get called on hitting the data limit almost immediately and the script keeps waiting for 15 minute intervals. For instance, the first time I tried today, the first user's 5 tweets downloaded just fine but when it attempted downloading the second user, I got hit with a 15 minute wait period. I have a developer Twitter account and bearer, and am nowhere near my 10 million monthly quota. Very befuddled...




I also have the same issue. My first guess was my developer account might reached 10 million. But, we just tried the same with my collogues and they have the same issue. Below is my get all tweets query

```

yo_tweets <- get_all_tweets(
  query = happy_emoji,
  start_tweets = "2016-01-01T00:00:00Z",
  end_tweets = "2021-07-18T00:00:00Z",
  data_path = "happy",
  bind_tweets = F,
  n = Inf,
  bearer_token = bearer_token,
  country = "ng"
)
```

Below is the rate limit reached messaages. This has been happening for a couple of days. 

```

Total pages queried: 1 (tweets captured this page: 3).
Total pages queried: 2 (tweets captured this page: 1).
Total pages queried: 3 (tweets captured this page: 0).
Total pages queried: 4 (tweets captured this page: 0).
Total pages queried: 5 (tweets captured this page: 2).
Rate limit reached. Rate limit will reset at 2021-07-18 18:28:48 
Sleeping for 896 seconds. 

==============================================================
Total pages queried: 6 (tweets captured this page: 0).
Total pages queried: 7 (tweets captured this page: 0).
Total pages queried: 8 (tweets captured this page: 0).
Rate limit reached. Rate limit will reset at 2021-07-18 18:43:50 
Sleeping for 899 seconds. 
==============================================================Total pages queried: 9 (tweets captured this page: 0).
Total pages queried: 10 (tweets captured this page: 0).
Rate limit reached. Rate limit will reset at 2021-07-18 18:58:53 
Sleeping for 900 seconds. 
==============================================================Total pages queried: 11 (tweets captured this page: 0).
Total pages queried: 12 (tweets captured this page: 0).
Rate limit reached. Rate limit will reset at 2021-07-18 19:13:56 
Sleeping for 900 seconds. 
==============================================================Total pages queried: 13 (tweets captured this page: 0).
Total pages queried: 14 (tweets captured this page: 0).
Rate limit reached. Rate limit will reset at 2021-07-18 19:28:59 
Sleeping for 899 seconds. 
==============================================================Total pages queried: 15 (tweets captured this page: 0).
Total pages queried: 16 (tweets captured this page: 0).
Total pages queried: 17 (tweets captured this page: 0).
Rate limit reached. Rate limit will reset at 2021-07-18 19:44:01 
Sleeping for 899 seconds. 

```
Please, kindly advise.

Thanks for the good work to the community @justinchuntingho @cjbarrie @chainsawriot 

Best,
Shamsuddeen

**Reply by cjbarrie** [reply]
Could you tell us what the `happy_emoji` object is referring to?

**Reply by cjbarrie** [reply]
My assumption is this: that there has been a change in how status codes are returned on the Twitter side and it is sending a 429 error when no tweets are returned for given time units... This is a bit strange and I'm not sure we can rectify it on our side. I have recreated this problem by using unicode "happy face," which is what I assume you are doing.

It seems to happen after 0 tweets are returned in quick succession. This might therefore require us introducing some delay in the API calls to avoid 429 status codes being returned. Will look into it

**Reply by shmuhammadd** [reply]
Thank you for the response. Below is what happy emoji is:

```
happy_emoji = c("àti 😃", "àti 😄", "àti 😆",
                   "àti 😊", "àti 😍", "àti 😗", "àti 😘", "àti 😙", "àti 😚", "àwọn 😃",
                   "àwọn 😄", "àwọn 😆", "àwọn 😊", "àwọn 😍", "àwọn 😗", "àwọn 😘", "àwọn 😙",
                   "àwọn 😚", "bá 😃", "bá 😄", "bá 😆", "bá 😊", "bá 😍", "bá 😗", "bá 😘", "bá 😙",
                   "bá 😚", "bí 😃", "bí 😄", "bí 😆", "bí 😊", "bí 😍", "bí 😗", "bí 😘", "bí 😙",
                   "bí 😚", "fi 😃", "fi 😄", "fi 😆", "fi 😊", "fi 😍", "fi 😗", "fi 😘", "fi 😙",
                   "fi 😚", "fún 😃", "fún 😄", "fún 😆", "fún 😊", "fún 😍", "fún 😗", "fún 😘",
                   "fún 😙", "fún 😚", "jẹ́ 😃", "jẹ́ 😄", "jẹ́ 😆", "jẹ́ 😊", "jẹ́ 😍", "jẹ́ 😗",
                   "jẹ́ 😘", "jẹ́ 😙", "jẹ́ 😚",  "kí 😃", "kí 😄", "kí 😆", "kí 😊", "kí 😍", "kí 😗",
                   "kí 😘", "kí 😙", "kí 😚", "kò 😃", "kò 😄", "kò 😆", "kò 😊", "kò 😍", "kò 😗", "kò 😘",
                   "kò 😙", "kò 😚", "láti 😃", "láti 😄", "láti 😆", "láti 😊", "láti 😍", "láti 😗", "láti 😘",
                   "láti 😙", "láti 😚", "lè 😃", "lè 😄", "lè 😆", "lè 😊", "lè 😍", "lè 😗", "lè 😘", "lè 😙", "lè 😚")

```

**Reply by chainsawriot** [reply]
It is actually due to the [1 req per 1 sec limit of the fulltext search endpoint](https://developer.twitter.com/en/docs/twitter-api/rate-limits).

Usually, a request takes more than 1 sec. But for request with no result, it would take less than 1 sec.

Of course, the easiest way to deal with this is to add 1 sec sleep after an empty request. 

A smarter approach is to keep track of the previous request timestamp (t0) and then compare the current time (t1) with the previous request time. 

Delta := t1 - t0 < 1, 
If Delta < 1, add a sleep time of the Delta plus a small surplus (e.g. 10 microsecs); then make the new request.

**Reply by chainsawriot** [reply]
@shmuhammad2004 BTW, "àti 😃" means you want to have tweets containing both `àti` and 😃 right? If it's the case, please quote all of them, i.e. `(àti 😃)`. 

``` r
require(academictwitteR)
#> Loading required package: academictwitteR
happy_emoji <- c("(àti 😃)", "(àti 😄)", "(àti 😆)",
                "(àti 😊)", "(àti 😍)", "(àti 😗)")

build_query(happy_emoji)
#> [1] "((àti 😃) OR (àti 😄) OR (àti 😆) OR (àti 😊) OR (àti 😍) OR (àti 😗))"
```

<sup>Created on 2021-07-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Reply by chainsawriot** [reply]
Actually, if you format your query correctly, it should work even without a fix.

**Reply by John-Kaz** [reply]
> It is actually due to the [1 req per 1 sec limit of the fulltext search endpoint](https://developer.twitter.com/en/docs/twitter-api/rate-limits).
> 
> Usually, a request takes more than 1 sec. But for request with no result, it would take less than 1 sec.
> 
> Of course, the easiest way to deal with this is to add 1 sec sleep after an empty request.
> 
> A smarter approach is to keep track of the previous request timestamp (t0) and then compare the current time (t1) with the previous request time.
> 
> Delta := t1 - t0 < 1,
> If Delta < 1, add a sleep time of the Delta plus a small surplus (e.g. 10 microsecs); then make the new request.

Thanks @chainsawriot! I added a 1 second wait time between iterations of the for loop, and this seems to have solved the problem.

**Reply by chainsawriot** [reply]
I just want to let you know that the current version 70e11bb67b1744496a0c02d05115eb7c263d4c5a has patched this.
Please consider updating.

**Reply by shmuhammadd** [reply]
@chainsawriot Thanks a lot. We appreciate you all for this project.
